distributed/comm/asyncio_tcp.py

```diff
@@ -0,0 +1,782 @@
+            try:
+                self._abort()
+            except RuntimeError:
+                # This happens if the event loop is already closed
+                pass
+
+    async def _close(self):
+        if self._transport is not None:
+            self._transport, transport = None, self._transport
+            transport.close()
+        await self._is_closed
+
+    def connection_made(self, transport):
+        if type(transport) is asyncio.selector_events._SelectorSocketTransport:
+            transport = _ZeroCopyWriter(transport)
+        self._transport = transport
+        if self.on_connection is not None:
+            self.on_connection(self)
+
+    def get_buffer(self, sizehint):
+        if self._frames is None or self._frame_nbytes_needed < self._default_len:
+            self._using_default_buffer = True
+            return self._default_buffer[self._default_end :]
+        else:
+            self._using_default_buffer = False
+            frame = self._frames[self._frame_index]
+            return frame[-self._frame_nbytes_needed :]
+
+    def buffer_updated(self, nbytes):
+        if nbytes == 0:
+            return
+
+        if self._using_default_buffer:
+            self._default_end += nbytes
+            while self._parse_next():
+                pass
+            self._reset_default_buffer()
+        else:
+            self._frame_nbytes_needed -= nbytes
+            if not self._frames_check_remaining():
+                self._message_completed()
+
+    def _parse_next(self):
+        if self._nframes is None:
+            if not self._parse_nframes():
+                return False
+        if len(self._frame_lengths) < self._nframes:
+            if not self._parse_frame_lengths():
+                return False
+        return self._parse_frames()
+
+    def _parse_nframes(self):
+        # TODO: we drop the message total size prefix (sent as part of the
+        # tornado-based tcp implementation), as it's not needed. If we ever
+        # drop that prefix entirely, we can adjust this code (change 16 -> 8
+        # and 8 -> 0).
+        if self._default_end - self._default_start >= 16:
+            self._nframes = struct.unpack_from(
+                "<Q", self._default_buffer, offset=self._default_start + 8
+            )[0]
+            self._default_start += 16
+            self._frame_lengths = []
+            return True
+        return False
+
+    def _parse_frame_lengths(self):
+        needed = self._nframes - len(self._frame_lengths)
+        available = (self._default_end - self._default_start) // 8
+        n_read = min(available, needed)
+        self._frame_lengths.extend(
+            struct.unpack_from(
+                f"<{n_read}Q", self._default_buffer, offset=self._default_start
+            )
+        )
+        self._default_start += 8 * n_read
+
+        if n_read == needed:
+            self._frames = [memoryview(bytearray(n)) for n in self._frame_lengths]
```

Comment:
See the end of the comment here: https://github.com/dask/distributed/pull/5450#issuecomment-962056466

> I'm wondering which is more common, and which we should optimize for right now. #5107 makes me think that lots of small frames is unfortunately the common case currently. As I understand it, all of these frames do belong to the same logical message object that was passed into `comm.write()`. Of course, worker data transfer also tries to ensure that every `get_data` message contains multiple pieces of disparate data. So really, we're pretty much guaranteed to run into both problems right now. I'm not sure which is worse.

All frames are from the same object passed to `comm.write` (since we're only parsing one message at a time, this is by definition true). However, they might not all be from the same logical python object, or have the same object lifetime. For example, suppose the message is a list of numpy arrays - this will be broken down into a frame for each array (or more frames if frame splitting is in effect), plus some header information. If we allocate a single large buffer for all frames, then holding on to any sub-frame will keep the whole buffer around (which is bad). I'd rather continue to do the right thing here and fix frame splitting elsewhere later.